lib/client.js

```diff
@@ -655,8 +659,14 @@ MqttClient.prototype._cleanUp = function (forced, done) {
  */
 MqttClient.prototype._sendPacket = function (packet, cb) {
   if (!this.connected) {
-    if (packet.qos > 0 || packet.cmd !== 'publish' || this.queueQoSZero) {
+    if ((packet.qos === 0 && this.queueQoSZero) || packet.cmd !== 'publish') {
       this.queue.push({ packet: packet, cb: cb })
+    } else if (packet.qos > 0) {
+      storePacket(this, packet, function (err) {
+        if (err) {
+          return cb && cb(err)
+        }
+      })
```

Comment:
you can just avoid the `storePacket` function, and use `client.outgoingStore.put` here.